### ISSUE TITLE
How to ace the Senior / Staff Engineer Interviews

### ISSUE THREAD

**Issue by stopachka** [OneGraph/essay.dev#42]
My last job search lasted 100 days. I went through 11 onsites at companies like Google, Airbnb, Lyft, Uber, Amazon, Square, Stripe — you name it and I did an onsite there. Out of the 11 I landed 10 offers, mostly as Staff Engineer, ultimately choosing Airbnb. 

The process changes drastically as you move into the Senior / Staff levels. I wanted to write this essay to draw an outline of the kind of preparation and change in focus that's needed and to do well in these interviews. I've seen little written about this, and I hope it's useful to you. This essay is formatted like a "100 day curriculum" of sorts. I hope you enjoy it! 

# Day 0: Foundations

To kick off the search, there are a few mindset shifts that if done well, can make everything else simple. In fact, I’ll venture to say that some of these mindset shifts when generalized can change your perspective… on life itself!

Okay okay that’s a serious statement, but give this a read, try it yourself and let me know what you think after 😊.

## **Foundations I: Attitude**

Most people _dread_ the job search. They think it’s a kind of performance, and _“they just want it to be over with”._ This kind of attitude can shoot you in the foot. If you dread something, you’re less likely to invest the time needed to get _great_ at it. That can produce negative experiences, which reinforces your opinion and causes a spiral of bad outcomes. Plus, if you _“just want it be over with”_ you’re more likely to choose the first few offers and miss out on the opportunity that’s _really_ right  for you.

Let’s shift this. **Learn to love the job search_._** You can look it as an infrequent, leveraged opportunity for you to grow. This is your chance to select _exactly_ the right next move for your career. It’s one of the best times to negotiate your role and your compensation. Plus, it’s a great opportunity for you to expand and deepen your relationships: you’ll lean on your community throughout the search and meet many talented people at some amazing companies.

Well, that’s a lot of real good reasons to love the job search. Internalize this attitude, and every step that follows turns into a meaningful part of your life story.

## **Foundations II: Community**

Most people treat the job search as a solo activity. They apply for jobs by themselves, they negotiate themselves, and they choose themselves. I think this is because people associate asking for help with _dependance,_ and they’re afraid to make themselves vulnerable.

Now, this is fine, it’s an _independent mindset._ But we can go further. Let’s **shift to an interdependent mindset_._** Think about it, teams achieve great things, not individuals_._

**Form a team around your search** and involve them in every phase**.** Who are the mentors who can show you what you didn’t know you didn’t know? Who are the peers that can train you and give you advice? Who are friends who can refer you to the right companies? Engage them from the start.

For example, before starting the search I met up with friends and mentors to get their advice on what to do. I can honestly tell you that these conversations opened paths I could never have _conceived_ before. **People who champion you can give you ambitious advice that you may be afraid to see**, and that changes the game**.** A special shoutout to my previous co-founder and life-long friend Sebastian Marshall, whose strategic advice changed was invaluable to me at this stage.

When I did start the search, I created a document that tracked the companies I was aiming for, alongside todos. This made it easy to share where I was at with my community, and easy for my community to discover ways they can help. Through this I received referrals at all 11 companies I aimed for.

When I began, two friends took the time to write up detailed advice from their last job search. Many, many friends took the time to train me. One of them was 12 timezones away and made time to skype for hours. They showed me where the bar was at, and gave me the feedback I needed to iterate.

![](https://github.com/stopachka/stopablog/blob/master/img/704/0*xmSh0y9yFwfcy8CK.png?raw=true "Had 1 day in New York, 3 life-long friends took the time to give me personal training")

Abraham Sorock, Alex Reichert, Andrew Wong, Daniel Woelfel, Giff Huang, Jacky Wang, Kam Leung, Mark Shlick, Robert Honsby, Sean Grove special shoutouts for the support and advice you gave me 😊. There’s many more — if you’re reading this and I haven’t included you, it’s mainly because I thought you preferred privacy — let me know and will update!

You’re starting to get a sense _how_ interdependent this was. If I listed every person, I think it would surpass a hundred people. I truly believe if you embrace this mindset and look at problems as a _team,_ you’ll see a huge change in all you go for, not just the search.

## **Foundations III: Narrative**

We get to the final part of the mindset, _how you look at your career._ Most people, if you ask them “_Why are you looking for a new job”,_ will give you either a generic answer — “I’m looking for a change_”,_ “I want to be impactful”, or some negative story _—_ “I was bored_”, “\_I didn’t like X_”._ This hurts you in at least three ways. First, it’s what everybody says, and it demonstrates that you haven’t thought deeply about your career. Second, it makes it hard for people to help_._ If you tell your mentors “you’re looking to be impactful”, that doesn’t narrow much down, and They can’t just open up their rolodex to email every company they know. Third, it makes it impossible to be a \_strong fit_ for any company. What can a recruiter note down if you say “you’re looking for a change”, which will let them think you’re a perfect fit for their company? If you’re looking for _any company,_ then by definition you can’t be a _strong fit_ for a specific company.

Instead, let’s shift your viewpoint and take charge of your career. **Look at your career as an adventure.** Sit down and reflect: What got you here? What did you _love_ about your last job? What would the most _amazing_ opportunity look like? As you go deeper, you’ll discover reasons that will get your thrilled, and show you how unique and awesome your path has already been. It will make the search a lot more fun, it will make it easier for people to help, and get recruiters _very_ excited about you. Most importantly, it will change how you think about yourself and your career. Let me be clear, this isn’t some trick to change your viewpoint artificially. If you dig deep, you’ll discover _real,_ true reasons, and that will reverberate throughout your life.

# Day 1 → 5: Kick Off

Okay, now we’ve learned to love the job search, we see it as a team endeavor, and we have a strong conviction about our career. It’s time to kick off the search. To do that, let’s do two things:

## **Kick Off I: Communication**

In every interview you do, people aim to extract _signal_ and _scope_. **Signal** is what _demonstrates_ that you can do your job — from coding to designing systems. **Scope** is what demonstrates _the kinds of problems you can take on —_ do you lead a team, a team of teams, or the whole org?

The combination of signal and scope is what determines your level_._ Your level, and where you fall on the spectrum is the most leveraged decision for determining your compensation and the kinds expectations you have.

Many people get annoyed about this. It feels very bureaucratic. Why should some number define you? You may also know people who have “high levels” but produce less. Thinking this way can hinder you. Another way you can look at it is this: it’s how things are for now, and it’s the best way we know so far to align your skills with the company. Let’s lean in and work with it.

Your goal for the first 5 days is to **form a conviction about your level.** Go over expectations, and prove to yourself that you are the level you are. If you want to do go deeper on this, visit our course and see the “Communication” module — it gives you a leveling guide to calibrate. Once you’ve done this, update your resume and LinkedIn to communicate that level. When you speak with recruiters and interviewers, you should be able to freely communicate it as well. Finally, go to levels.fyi, and form a conviction about the kind of compensation ranges you’re looking for too.

## **Kick Off II: Schedule Screens**

From consulting with your community and forming your narrative, you should have about 10 or so companies that you could see yourself being _thrilled_ to work at_._ Now, engage your community and find a referral for each company. You can do this by searching through your LinkedIn, or sharing a document of your top choices with your community and asking them if they know people who can refer. Do the best you can to go through referrals, as having someone to champion you helps in many ways throughout the process. If you can’t find someone, it’s okay to send a direct application, but do that as a very last resort.

With all recruiters, ask to schedule a screen about 20–30 days in advance. They may be surprised, but you can tell them you want to prepare, and that will send positive signal too. With that, you’ll have a deadline, and a real good reason to start studying 😊.

# Day 5 → 30: Preparation

Now comes the hard work. It’s time to ramp up for the technical interviews.

## **Preparation I: Key Mindset**

The mindset at this stage I want to stress, is to _give it all you’ve got._ Sometimes, as a way of coping with rejection, people give themselves an out — “Oh I didn’t really study anyways”. My suggestion is to **frame your success on how hard you prepare,** not on the results. Aim to remove _any reason to_ say, you didn’t try your absolute best.

Fun story on that — months before the search, I had trip planned with one of my closest friends to go to Israel and Jordan, and it fell right during preparation time…

![](https://github.com/stopachka/stopablog/blob/master/img/704/0*7sqMOAxWpQV5TiYy.png?raw=true "Elements of Programming Interviews, 6am at a Bedouin camp")

Well then, we need to travel and study at the same time 🙂. A special shoutout for Jacky Wang for being such an awesome travel partner, and being an encourager for the studies every day. Jacky’s the kind of guy who not only pushes the envelope to explore (just check out his [photos](http://jackywang.org)), but he’ll indulge and converse with excitement about programming questions _for hours_ on long car rides!

## **Preparation II: General Advice**

There are three pieces of advice that apply generally to all interview types. First, the process to study: I recommend self-study first, then schedule practice interviews, then try the real thing, then calibrate and repeat the process if necessary. Second, **approach every interview as if you are solving the problem for real_._** Most people seem to “pretend” when they solve the problem. They may not go deep enough, write code that won’t run, explain things for the sake of explaining, or create a system that won’t scale. _Imagine a team at work set up a meeting and asked you to solve the problem,_ and act accordingly. Third, most people are blissfully unaware of what an _excellent performance_ looks like. They assume the interviewer will read their minds and see how amazing they are. As you can guess, this won’t work. Study each interview type deeply, and get feedback on how you’re doing from your peers. Calibrate, make sure you are aware the bar, and you are well above it.

_…**Now, for some more specific advice**_

## **Preparation III: Algorithm Interviews**

Unless you’re a specialist, this is the interview type to master  first. No matter how skilled and senior you are, algorithm interviews play a significant role. I suggest you pick one book — Joe and I prefer [Elements of Programming Interviews](https://www.amazon.com/Elements-Programming-Interviews-Python-Insiders/dp/1537713949/) — and work through it linearly every day (personal nit: skip the first chapter in this book on bitwise operators, this is the most annoying part of the book, and isn’t needed for most interviews). If you get stuck on a question, just look at the answer, but make sure to write it with your own code, line by line. By the end of it, you’ll be surprised with all that you learn, and I guarantee you, at least if you go through this book, you’ll rarely be asked a question that you can’t map to something you’ve learned.

One common pitfall here is, people go on leetcode and do a bunch of random questions. I suggest avoiding this — since the questions are random, you won’t see improvement quickly, and you may not learn what’s needed. Leetcode is great though for finding and doing specific questions.

A week or so after you’re into the book, schedule practice interviews. My favorites are <https://interviewing.io> and <https://pramp.com>. A special shoutout to the founder of interviewing.io, Aline Lerner. There was an obscure 19 year-old on hackernews who was having trouble coming to the U.S. She took the time to help, and even gave him free access to her lawyers. That 19 year-old was me, and I am still grateful 🙂.

The third or fourth week, schedule practice interviews with your friend group. This is where you’ll _really_ get a sense of where you’re at. If the feedback is positive, you’re ready for the screens. If not, \_move them. Y_ou’ll be surprised how many times you can do that. We go quite deep on this, and the kinds of questions you should be able to answer, in the “Algorithm Interview” module of jobsearch.dev.

## **Preparation IV: UI Interviews**

For UI Engineers, I suggest three steps. First, get comfortable with online editors, you’ll be using them all the time, so you should be able to set up an environment quickly. Second, ramp back up on some UI concepts that you may not have worked on recently — things like drag and drop, selection apis, etc. If you’re looking for more specific inspiration, we’ve included some homework in the “UI Interview” module of the course. Third, explore the latest and greatest in JS — look up what interests you — from react hooks to all that’s new in GraphQL. If you’re looking for some courses, [Wes Bos](https://wesbos.com) has some great content to ramp you up. The key to these interviews is to demonstrate your ability to build UIs and your love of UI engineering. If you follow these steps it’ll come through naturally.

## **Preparation V**: **System Design Interviews**

For System Design, I also suggest three steps. First, master the _structure_ — scope down, go broad, outline a complete solution, go deep on as many components as you can, then go the extra mile, to share how you would actually launch the system. The “System Design” module in the course does a great job of going deep on this. Next, as you go through the practice interviews, you’ll start to notice places where you can improve. Start noting those down and build a plan to learn. For example, if you don’t know how to scale, you can search “{Company Name} InfoQ” and find some awesome talks. If you’re unsure about concurrency, the book [7 concurrency models in 7 weeks](https://www.amazon.com/Seven-Concurrency-Models-Weeks-Programmers/dp/1937785653) can ramp you up. Don’t know much about databases? [7 databases in 7 weeks](https://www.amazon.com/Seven-Databases-Weeks-Modern-Movement/dp/1680502530). Pick what interests you and start going deep. This is my favorite interview type, because as you go deeper, the skills and knowledge you learn transfers into making you a better architect.

## **Preparation VI: Experience Interviews**

The purpose of the Experience Interview is to understand your _scope —_ the kinds of problems you can solve — and whether you are a culture fit. This interview type largely takes care of itself if you’ve centered on your narrative (you know what you want and where you’re going), and your communication (you know what level you are). The first coneys your culture fit, and the second your scope. You can go deeper on this in the “Experience Interview” module in the course.

## **Preparation VII: How to Focus for these 20 days**

Now you’ve got an outline of each kind of interview, and how to prepare. For the first 20 days, I suggest focusing heavily on Algorithms. If you’re a UI engineer, still do algos, but spend maybe 30–40% of the time building UI components too. You want to focus on this because the screens are mainly this type of interview. As you pass the screens, you’ll have to go deeper on systems before you head into the onsites.

# Day 30 → 60: Execution

Okay, we’re now ready to execute. Over the next month, you should be finished with all your screens and on-sites.

## **Execution I: Key Pieces**

Three key pieces of advice here:

Come in to every interview with a positive mindset, and **treat everyone with class**. These people could become your friends, and it’s fun to learn about how a company works. Come with that mindset, and the confidence will seep through.

_Batch all the interviews._ **Make sure all the screens and onsites are scheduled around the same time**. The preparation you can do part time, but this piece Joe and I highly suggest, you figure out how to do full-time. When offers land around the same time, you’ll have a lot more leverage.

**Communicate your level and your narrative**_._ Keep signal and scope top of mind, and communicate clearly during every phase — recruiter screens, technical screens, and and onsites. We go deeper on this in the “Interview Phases” module of the course.

## **Execution II: Dealing with the Unexpected**

As you go through the onsites, two things can happen:

**You’ll fail at some interview**

As you may have noted, I did 11 onsites, and got 10 offers…that means I failed one of em 😄. _Even with some of the most intense prep, sh\*t happens_. When it does, do not over-react, embarrass yourself by being mean, or change your plans and cancel all upcoming interviews_._ Instead, treat everyone with a smile and with class. Then, be kind to yourself, take out what lessons you can, and _continue on with the plan_. If you see failure on more than one or two interviews, consider re-scheduling and re-calibrating.

**You’ll get something unexpected**

Life can throw curve ball at anytime, and you’re unlikely to avoid it during the job search. Case in point:

![](https://github.com/stopachka/stopablog/blob/master/img/704/0*QdooVDcBLBCDc-De.png?raw=true "Don’t try huge jumps on your skis, when you don’t know how, especially during your job search")

On Christmas break, Joe and I went skiing. I did an unintended backflip, and landed in the ER. When something like this happens, again, avoid over-reacting. Treat surprises as an opportunity to be malleable, and make it part of your story. For example, the day I visited the ER, I had a conversation with one of the best recruiters I’ve ever met, Twitter’s Matt Robbins. I sent him that picture, and I wager that will be the most unique recruiter conversation either of us will have in our career.

# Day 60 → 100: Choice

Around Day 60, offers should start to roll in, and it’s time to choose. The first surprise that people may have here, is the number of days. Most people take about a week or two to decide…Day 60–100 is 40 days! Does it really take that long to choose a company, even after you’ve received your offers?

Yes!

Let me be clear, this isn’t about negotiation. You’ve already done 90% of the legwork for that through your preparation and execution — if you communicated your level well, you’ve aced the interviews, and you landed your offers around the same time, you can safely assume every company will be competitive and get you the best offer. There are specific things you can do at this point — the main advice being to use clear with what you would sign right away with. We go deeper on this in the “Negotiation” section of the course. Patio11’s [Salary Negotiation](https://www.kalzumeus.com/2012/01/23/salary-negotiation/) and Rands’ [The Business](http://randsinrepose.com/archives/the-business/) are great reading too.

The reason this takes so long is because the purpose for this phase is to **_choose the company that’s the best fit for you._** It’s your turn to interview back. Meet your team, meet your manager, really align on the kinds of work you would do, and let your future team show you how they could be a part of your narrative.

To give you context, Stripe, Airbnb, Square, and Google were my final choices. I ended up meeting 6 managers at google, and visited Stripe, Airbnb and Square offices at least 7 times each. By the end of these meetings, you’ll have formed friends, and have gotten a _great_ sense of what it’s like to work at each company.

The final step before making the decision, is to get feedback from your community. I suggest you **write out an essay that goes over your decision,** and discuss with your peers and mentors — do they agree with your thinking, what do they see that you don’t, do they have any suggestions?

To get a sense of how this looks, here’s a compressed screenshot of the essay I sent my friends and mentors: “On Choosing”

![](https://github.com/stopachka/stopablog/blob/master/img/704/0*mM8hJ6aAcYH9nV39.png?raw=true "Don’t squint, it’s not meant to be read")

It’s a bit too personal to share outright, but I wanted to give you a sense of the kind of deliberation that went behind this: thousands of words, and insights from more than a dozen of my closest friends and mentors. These were amazing companies and the decision was close.

If you do this well, you’ll have clear, strong conclusions for _exactly_ why you would choose a certain company. Compensation will be far away from the main reasons, and you’ll be thrilled to sign.

# Day 100: Breathe

This will be a day where you breathe and realize how exciting, yet how tiring things were. Take a break and plan a celebration with your community.

**With that, we’ve covered the advice 😊.**

# Final Notes

## Want to go deeper?

If you made it this far, and you’re pumped, get ready to go deep into the rabbit hole. Joe, Luba, and I, with the help of Aamir Patel and Jordan Yagiello, expounded on all the lessons learned here, and created a video course that will walk you through it all. From leveling guides, to example narratives, to deep explanations on system design, UI interviews, and more, you can get it there. **Best of all, it’s 100% free.** We were thinking of making it paid course initially, but after some long talks, we wanted to share this with our community 😊. Go to <https://jobsearch.dev> and see for yourself.

## Onwards from Airbnb

My best friend and co-founder Joe and I left Airbnb, to start a company of our own. We are currently exploring fitness, figuring out what it takes to reach advanced levels of performance. If that kind of stuff interests you, sign up for our [waitlist](https://consistent.fit) here.
